src/backend/distributed/commands/utility_hook.c

```diff
@@ -644,6 +666,180 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
 	}
+
+	if (list_length(ddlJobs) != 0)
+	{
+		return true;
+	}
+
+	if (IsDropSchemaOrTable(parsetree))
+	{
+		/*
+		 * TODO: We could specifically check if it is a:
+		 * - DROP TABLE statement that drops a reference or a citus local table
+		 *   that has a foreign key relationship with a postgres table, or a
+		 * - DROP SCHEMA statement that again drops such a citus table
+		 * but this is more easy and the other approach is much more prune to
+		 * deadlocks.
+		 */
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * IsDropSchemaOrTable returns true if parsetree represents a DROP SCHEMA
+ * or a DROP TABLE statement.
+ */
+static bool
+IsDropSchemaOrTable(Node *parsetree)
+{
+	if (!IsA(parsetree, DropStmt))
+	{
+		return false;
+	}
+
+	DropStmt *dropStatement = (DropStmt *) parsetree;
+	ObjectType dropType = dropStatement->removeType;
+	return dropType == OBJECT_SCHEMA || dropType == OBJECT_TABLE;
+}
+
+
+/*
+ * UndistributeCitusLocalTablesIfNeeded undistributes citus local tables that
+ * are not connected to any reference tables via their individual foreign key
+ * subgraphs.
+ */
+static void
+UndistributeCitusLocalTablesIfNeeded(bool executedDDLJob)
+{
+	if (!ShouldUndistributeCitusLocalTables(executedDDLJob))
+	{
+		return;
+	}
+
+	List *citusLocalTableIdList = CitusTableTypeIdList(CITUS_LOCAL_TABLE);
+	citusLocalTableIdList = SortList(citusLocalTableIdList, CompareOids);
+
+	/* acquire ExclusiveLock as UndistributeTable does */
+	LockRelationsWithLockMode(citusLocalTableIdList, ExclusiveLock);
+
+	Oid citusLocalTableId = InvalidOid;
+	foreach_oid(citusLocalTableId, citusLocalTableIdList)
+	{
+		HeapTuple heapTuple =
+			SearchSysCache1(RELOID, ObjectIdGetDatum(citusLocalTableId));
+		if (!HeapTupleIsValid(heapTuple))
+		{
+			/*
+			 * UndistributeTable drops relation, skip if already undistributed
+			 * via cascade.
+			 */
+			continue;
+		}
+		ReleaseSysCache(heapTuple);
+
+		if (ConnectedToReferenceTableViaFKey(citusLocalTableId))
+		{
+			/* still connected to a reference table, skip it */
+			continue;
+		}
+
+		/*
+		 * Citus local table is not connected to any reference table, then
+		 * undistribute it via cascade.
+		 */
+		TableConversionParameters params = {
+			.relationId = citusLocalTableId,
+			.cascadeViaForeignKeys = true
+		};
+		UndistributeTable(&params);
+	}
+}
+
+
+/*
+ * ShouldUndistributeCitusLocalTables returns true if we might need to check
+ * citus local tables for their connectivity to reference tables.
+ */
+static bool
+ShouldUndistributeCitusLocalTables(bool executedDDLJob)
+{
+	if (!ConstraintDropped)
+	{
+		/*
+		 * citus_drop_trigger executes notify_constraint_dropped to set
+		 * ConstraintDropped to true, which means that last command dropped
+		 * a table constraint.
+		 */
+		return false;
+	}
+
+	if (UtilityHookLevel != 1)
+	{
+		/* we only perform this operation for top level DDL command */
+		return false;
+	}
+
+	ResetConstraintDropped();
+
+	if (!EnableLocalReferenceForeignKeys)
```

Comment:
it is better to use `ShouldEnableLocalReferenceForeignKeys`

Reply:
1c7df53f499ed6c6cd81a3207f9269eca565f9f5